### PR TITLE
CovarianceBlock and associated MetaData

### PR DESCRIPTION
**Pull request by whaeck**
This PR introduces the CovarianceBlock and its associated MetaData. 

In general, we would have covariance data given as follows:
```
      N1,R1 N1,R2 N2,R1 ...
N1,R1   A     B     C    ...
N1,R2   B^T   D     E    ...
N2,R1   C^T   E^T   F    ...
...
```

A CovarianceBlock represents a piece of this data, corresponding to a single one of these matrices A, B, C, etc. The MetaData contains the information associated to the rows (and columns) of the block in question. The meta data is currently limited to the following:
- the nuclide and reaction the rows and columns are associated to
- the energy structure of the rows and columns

For a block on the diagonal (row and column metadata are the same) only the row metadata is stored internally (using std::optional). Such a diagonal block will be inherently represented by a square matrix. Off-diagonal blocks do not necessarily have to be square.

As such, these objects are currently only usable for covariance data directly associated to reaction data (i.e. MF31, MF33). While MF34 and MF35 data will fit in these, we'll need to expand the meta data object to contain additional information (i.e. the MF35 data needs the incident energy group for the covariance block). Resonance covariance data is another beast altogether and does not fit in these objects. Because of this, we may want to rename the CovarianceBlock to ReactionCovarianceBlock instead and create a ResonanceCovarianceBlock for resonance parameters themselves. 

Current shortcomings:
- we do not store whether or not the covariances are relative or absolute
- for things like angular distribution covariances, we need to store the reference frame

### CODE REVIEW

File: src/codex/MetaData/test/MetaData.test.cpp
Comment (ptalou) [inline src/codex/MetaData/test/MetaData.test.cpp:1]: When throwing an exception after testing the energies for identical adjacent values, could you put out the energies in the message?

File: src/codex/MetaData/src/verifyEnergies.hpp
Comment (ptalou) [inline src/codex/MetaData/src/verifyEnergies.hpp:1]: This function should be made static.

File: src/codex/MetaData/test/MetaData.test.cpp [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+        CHECK_THROWS( MetaData( std::move( nuclide ), std::move( reaction ),
+                                std::move( wrong ) ) );
+      } // THEN
+    } // WHEN
+
+    WHEN( "the energy grid has duplicate points" ) {
+
+      NuclideID nuclide( "U235" );
+      ReactionID reaction( "elastic" );
+      std::vector< double > wrong = { 2e+7, 1., 1., 1e-5 };
```
Comment (ptalou) [inline src/codex/MetaData/test/MetaData.test.cpp]: Those energies are not sorted, so the test will not check the adjacent values but fail before.

File: src/codex/CovarianceBlock/src/verifyMatrix.hpp [hunk 2]
```diff
@@ -0,0 +1,47 @@
+void verifyMatrix( const Matrix< double >& matrix,
```
Comment (ptalou) [inline src/codex/CovarianceBlock/src/verifyMatrix.hpp:1]: Even in the case of a square matrix block, on the diagonal of the full matrix, one needs to check that this matrix block is symmetric.

File: src/codex/CovarianceBlock/src/verifyMatrix.hpp [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    Log::info( "Number of energy groups: {}", energies.size() - 1 );
+    throw std::exception();
+  }
+}
+
+void verifyMatrix( const Matrix< double >& matrix,
+                   const std::vector< double >& rowEnergies,
+                   const std::vector< double >& columnEnergies ) {
+
+  // check if the matrix row order is consistent with the energy boundaries
```
Comment (ptalou) [inline src/codex/CovarianceBlock/src/verifyMatrix.hpp]: Same error message. Could be combined or specific for rows and columns, respectively.

File: src/codex/CovarianceBlock/src/ctor.hpp [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+                             std::move( energies ) ),
+                   std::move( matrix ) ) {}
+
+/**
+ *  @brief Constructor for an off-diagonal covariance block
+ *
+ *  @param[in] metadata   the metadata associated with the covariance block
+ *  @param[in] matrix     the covariance matrix
+ */
+CovarianceBlock( MetaData row, MetaData column, Matrix< double > matrix ) :
```
Comment (ptalou) [inline src/codex/CovarianceBlock/src/ctor.hpp]: I don't like the labels 'row' and 'column' for the metadata... maybe 'metadata_row'?

File: src/codex/CovarianceBlock/src/ctor.hpp [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+CovarianceBlock( NuclideID nuclide, ReactionID reaction,
+                 std::vector< double > energies, Matrix< double > matrix ) :
+  CovarianceBlock( MetaData( std::move( nuclide ), std::move( reaction ),
+                             std::move( energies ) ),
+                   std::move( matrix ) ) {}
+
+/**
+ *  @brief Constructor for an off-diagonal covariance block
+ *
+ *  @param[in] metadata   the metadata associated with the covariance block
```
Comment (ptalou) [inline src/codex/CovarianceBlock/src/ctor.hpp]: The documentation does not correspond to the arguments (copy/paste?)

File: src/codex/CovarianceBlock/test/CovarianceBlock.test.cpp [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+                                       std::move( rowEnergies ),
+                                       std::move( columnNuclide ),
+                                       std::move( columnReaction ),
+                                       std::move( columnEnergies ),
+                                       std::move( matrix ) ) );
+      } // THEN
+    } // WHEN
+
+    WHEN( "the matrix order is not consistent with the energy boundaries "
+          "for an off-diagonal covariance block (rows)" ) {
```
Comment (ptalou) [inline src/codex/CovarianceBlock/test/CovarianceBlock.test.cpp]: replace "covariance block (rows)" by "covariance block (columns)"

File: python/src/CovarianceBlock.python.cpp [hunk 7]
```diff
@@ -0,0 +1,96 @@
...
+    "    row_reaction      the row reaction identifier\n"
+    "    row_energies      the row energy boundaries\n"
+    "    column_nuclide    the row nuclide identifier\n"
+    "    column_reaction   the row reaction identifier\n"
+    "    column_energies   the row energy boundaries\n"
+    "    matrix            the covariance matrix"
+  )
+  .def_property_readonly(
+
+    "row",
```
Comment (ptalou) [inline python/src/CovarianceBlock.python.cpp]: Change to row_metadata?

File: python/src/CovarianceBlock.python.cpp [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+    "is_diagonal",
+    &Component::isDiagonal,
+    "Flag to indicate whether or not this covariance block is a diagonal block"
+  )
+  .def_property_readonly(
+
+    "matrix",
+    &Component::matrix,
+    "The covariance matrix",
+    python::return_value_policy::reference_internal
```
Comment (ptalou) [inline python/src/CovarianceBlock.python.cpp:97]: Maybe a word or two of explanation here?

**Review by ptalou — CHANGES_REQUESTED** [review]
Good work!

File: src/codex/MetaData/test/MetaData.test.cpp (hunk 1 above)
Comment (whaeck) [reply src/codex/MetaData/test/MetaData.test.cpp]: Good find. The Python test is correct though, weird.

File: src/codex/CovarianceBlock/src/verifyMatrix.hpp (hunk 2 above)
Comment (whaeck) [reply src/codex/CovarianceBlock/src/verifyMatrix.hpp:1]: Indeed. I actually added the symmetry test in a future PR so I will not address it here.

**Review by ptalou — APPROVED** [review]

**Pull request merged**